Lib/test/test_capi/test_complex.py

```diff
@@ -0,0 +1,161 @@
+
+        self.assertTrue(checkexact(1+2j))
+        self.assertFalse(checkexact(ComplexSubclass(1+2j)))
+        self.assertFalse(checkexact(Complex()))
+        self.assertFalse(checkexact(3))
+        self.assertFalse(checkexact(3.0))
+        self.assertFalse(checkexact(object()))
+
+        # CRASHES checkexact(NULL)
+
+    def test_fromccomplex(self):
+        # Test PyComplex_FromCComplex()
+        fromccomplex = _testcapi.complex_fromccomplex
+
+        self.assertEqual(fromccomplex(1+2j), 1.0+2.0j)
+
+    def test_fromdoubles(self):
+        # Test PyComplex_FromDoubles()
+        fromdoubles = _testcapi.complex_fromdoubles
+
+        self.assertEqual(fromdoubles(1.0, 2.0), 1.0+2.0j)
+
+    def test_realasdouble(self):
+        # Test PyComplex_RealAsDouble()
+        realasdouble = _testcapi.complex_realasdouble
+
+        # Test subclasses of complex/float
+        self.assertEqual(realasdouble(1+2j), 1.0)
+        self.assertEqual(realasdouble(-1+0j), -1.0)
+        self.assertEqual(realasdouble(-1.0), -1.0)
+
+        self.assertEqual(realasdouble(ComplexSubclass(1+2j)), 1.0)
+        self.assertEqual(realasdouble(3.14), 3.14)
+        self.assertEqual(realasdouble(FloatSubclass(3.14)), 3.14)
+
+        # Test types with __complex__ dunder method
+        # Function doesn't support classes with __complex__ dunder, see #109598
+        self.assertRaises(TypeError, realsasdouble, Complex())
+        #self.assertEqual(realasdouble(Complex()), 4.25)
+        #self.assertRaises(TypeError, realasdouble, BadComplex())
+        #with self.assertWarns(DeprecationWarning):
+        #    self.assertEqual(realasdouble(BadComplex2()), 4.25)
+        #with warnings.catch_warnings():
+        #    warnings.simplefilter("error", DeprecationWarning)
+        #    self.assertRaises(DeprecationWarning, realasdouble, BadComplex2())
+        #self.assertRaises(RuntimeError, realasdouble, BadComplex3())
+
+        # Test types with __float__ dunder method
+        self.assertEqual(realasdouble(Float()), 4.25)
+        self.assertRaises(TypeError, realasdouble, BadFloat())
+        with self.assertWarns(DeprecationWarning):
+            self.assertEqual(realasdouble(BadFloat2()), 4.25)
+
+        self.assertRaises(TypeError, realasdouble, 42)
+        self.assertRaises(TypeError, realasdouble, object())
+
+        # CRASHES realasdouble(NULL)
+
+    def test_imagasdouble(self):
+        # Test PyComplex_ImagAsDouble()
+        imagasdouble = _testcapi.complex_imagasdouble
+
+        # Test subclasses of complex/float
+        self.assertEqual(imagasdouble(1+2j), 2.0)
+        self.assertEqual(imagasdouble(1-1j), -1.0)
+        self.assertEqual(imagasdouble(ComplexSubclass(1+2j)), 2.0)
+        self.assertEqual(imagasdouble(3.14), 0.0)
+        self.assertEqual(imagasdouble(FloatSubclass(3.14)), 0.0)
+
+        # Test types with __complex__ dunder method
+        # Function doesn't support classes with __complex__ dunder, see #109598
```

Comment:
Add assertRaises and remove the rest of comment.

Reply:
It returns 0.0 here as well.  I've added an assert.